File: client-sdks/sockudo-kotlin/lib/src/test/kotlin/io/sockudo/client/PushRegistrationTest.kt
package io.sockudo.client

import kotlinx.coroutines.test.runTest
import okhttp3.mockwebserver.MockResponse
import okhttp3.mockwebserver.MockWebServer
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertTrue

class PushRegistrationTest {
    @Test
    fun publishUsesBackendProxyAndKeepsPushAsyncByDefault() =
        runTest {
            val server = MockWebServer()
            server.enqueue(
                MockResponse()
                    .setResponseCode(202)
                    .addHeader("Content-Type", "application/json")
                    .setBody("""{"publish_id":"pub_123"}"""),
            )
            server.use {
                val client =
                    SockudoPushRegistration(
                        PushRegistrationOptions(
                            endpoint = server.url("/push/").toString(),
                            headers = mapOf("Authorization" to "Bearer session"),
                        ),
                    )

                val response =
                    client.publish(
                        mapOf(
                            "recipients" to listOf(mapOf("type" to "channel", "channel" to "orders")),
                            "payload" to mapOf("title" to "Order", "body" to "Updated"),
                        ),
                    )

                val request = server.takeRequest()
                val body = JsonSupport.fromJsonElement(JsonSupport.decode(request.body.readUtf8())) as Map<String, Any?>

                assertEquals(mapOf("publish_id" to "pub_123"), response)
                assertEquals("/push/publish", request.path)
                assertEquals("POST", request.method)
                assertFalse(body["sync"] as Boolean)
                assertEquals("Bearer session", request.getHeader("Authorization"))
                assertTrue(request.getHeader("Content-Type")?.startsWith("application/json") == true)
            }
        }

    @Test
    fun updateDeviceRegistrationAddsDeviceIdentityTokenHeader() =
        runTest {
            val server = MockWebServer()
            server.enqueue(
                MockResponse()
                    .setResponseCode(201)
                    .addHeader("Content-Type", "application/json")
                    .setBody("""{"change":"updated"}"""),
            )
            server.use {
                val client =
                    SockudoPushRegistration(
                        PushRegistrationOptions(endpoint = server.url("/push/").toString()),
                    )

                client.updateDeviceRegistration(
                    device =
                        mapOf(
                            "id" to "device-1",
                            "formFactor" to "phone",
                            "platform" to "android",
                            "timezone" to "UTC",
                            "locale" to "en",
                            "push" to mapOf("recipient" to mapOf("transportType" to "gcm", "registrationToken" to "rotated")),
                        ),
                    deviceIdentityToken = "identity",
                )

                val request = server.takeRequest()
                assertEquals("/push/deviceRegistrations", request.path)
                assertEquals("identity", request.getHeader("X-Sockudo-Device-Identity-Token"))
            }
        }

    @Test
    fun listChannelSubscriptionsUsesCursorPaginationParams() =
        runTest {
            val server = MockWebServer()
            server.enqueue(
                MockResponse()
                    .setResponseCode(200)
                    .addHeader("Content-Type", "application/json")
                    .setBody("""{"items":[],"has_more":false}"""),
            )
            server.use {
                val client =
                    SockudoPushRegistration(
                        PushRegistrationOptions(endpoint = server.url("/push/").toString()),
                    )

                client.listChannelSubscriptions(
                    PushSubscriptionParams(
                        deviceId = "device-1",
                        limit = 10,
                        cursor = "c1",
                    ),
                )

                val request = server.takeRequest()
                assertEquals(
                    "/push/channelSubscriptions?deviceId=device-1&limit=10&cursor=c1",
                    request.path,
                )
            }
        }
}

File: client-sdks/sockudo-js/test/push-registration.test.ts
import { describe, expect, it } from "vitest";

import { SockudoPushRegistration } from "../src/core/push";

describe("SockudoPushRegistration", () => {
  it("uses a backend proxy and keeps push publish async by default", async () => {
    const requests: Array<{ url: string; init: RequestInit }> = [];
    const client = new SockudoPushRegistration({
      endpoint: "https://api.example.test/push",
      headers: { Authorization: "Bearer session" },
      fetch: (async (url: string, init: RequestInit) => {
        requests.push({ url, init });
        return new Response(JSON.stringify({ publish_id: "pub_123" }), {
          status: 202,
          headers: { "Content-Type": "application/json" },
        });
      }) as typeof fetch,
    });

    const response = await client.publish({
      recipients: [{ type: "channel", channel: "orders" }],
      payload: { title: "Order", body: "Updated" },
      providerOverrides: [{ provider: "fcm", payload: { android: {} } }],
    });

    expect(response).toEqual({ publish_id: "pub_123" });
    expect(requests[0].url).toBe("https://api.example.test/push/publish");
    expect(requests[0].init.method).toBe("POST");
    expect(JSON.parse(requests[0].init.body as string)).toMatchObject({
      sync: false,
      recipients: [{ type: "channel", channel: "orders" }],
    });
    expect(requests[0].init.headers).toMatchObject({
      Authorization: "Bearer session",
      "Content-Type": "application/json",
    });
  });

  it("passes device identity token only to device update requests", async () => {
    const requests: Array<{ url: string; init: RequestInit }> = [];
    const client = new SockudoPushRegistration({
      endpoint: "https://api.example.test/push",
      fetch: (async (url: string, init: RequestInit) => {
        requests.push({ url, init });
        return new Response(JSON.stringify({ change: "updated" }), {
          status: 201,
          headers: { "Content-Type": "application/json" },
        });
      }) as typeof fetch,
    });

    await client.updateDeviceRegistration(
      {
        id: "device-1",
        formFactor: "phone",
        platform: "android",
        timezone: "UTC",
        locale: "en",
        push: {
          recipient: { transportType: "gcm", registrationToken: "rotated" },
        },
      },
      "identity",
    );

    expect(requests[0].url).toBe("https://api.example.test/push/deviceRegistrations");
    expect(requests[0].init.headers).toMatchObject({
      "X-Sockudo-Device-Identity-Token": "identity",
    });
  });

  it("uses cursor pagination params on list calls", async () => {
    const urls: string[] = [];
    const client = new SockudoPushRegistration({
      endpoint: "https://api.example.test/push",
      fetch: (async (url: string, _init: RequestInit) => {
        urls.push(url);
        return new Response(JSON.stringify({ items: [], has_more: false }), {
          status: 200,
          headers: { "Content-Type": "application/json" },
        });
      }) as typeof fetch,
    });

    await client.listChannelSubscriptions({
      deviceId: "device-1",
      limit: 10,
      cursor: "c1",
    });

    expect(urls[0]).toBe(
      "https://api.example.test/push/channelSubscriptions?deviceId=device-1&limit=10&cursor=c1",
    );
  });
});
